Make Node a dataclass

adsharma opened the pull request:
Dataclasses make the code more compact and generate many convenience
functions.

adsharma commented on quoracle/quorum_system.py:
```diff
@@ -520,11 +520,17 @@ def fr_load(problem: pulp.LpProblem, fr: float) -> pulp.LpAffineExpression:
                     vs = x_to_read_quorum_vars[x]
-                    x_load += fr * sum(vs) / self.node(x).read_capacity
+                    xread_capacity =  self.node(x).read_capacity
+                    if xread_capacity is None:
```
This code would not be necessary if we made:

```
read_capacity: float = 1.0
```

in the dataclass. More instances below.

mwhittaker commented:
Ah, I've been targeting Python 3.6, and I think dataclasses are a 3.7+ thing?

adsharma commented:
Backports are available: https://pypi.org/project/dataclasses/

In general, dataclasses eliminate so much boilerplate code that I use them wherever I can.

mwhittaker commented on quoracle/expr.py:
```diff
@@ -89,14 +90,19 @@ def _dup_free_min_failures(self) -> int:
+    write_capacity: Optional[float] = None
+    latency: Optional[datetime.timedelta] = None
+
+    def post_init(self):
```
This should be `__post_init__`.

mwhittaker reviewed (commented):
Thanks for the PR! I spent some time reading up on dataclasses, and I think the switch here may not be worth it. 

- I don't think we benefit much since there's not a lot of boilerplate to remove in this case. 
- The dataclass also introduces `capacity` into `Node` which it didn't previously have, and it's not clear what `capacity` means. For example, if we construct a node like `Node('a', read_capacity=100, write_capacity=200)`, there's not a nice value for `capacity`.
- Also since `Node` is part of the `Expr` hierarchy, it feels weird to me that `Node` would be a dataclass but not `And` or `Or`.

adsharma commented:
Yes - it saves only 10 lines in this diff. Not an earth shattering difference :)

I've addressed the bugs you found in the code and simplified the typing and default values.

I'll leave it here just in case you end up using dataclasses elsewhere and then want to convert this one for consistency.
